### PR TITLE
Speed up get_linking_probabilities

### PR DESCRIPTION
**Pull request by matt-gardner**
This reduces the percent of computation done in `_get_linking_probabilities` to approximately zero, giving a roughly 30% reduction in training time.

@rajasagashe, github isn't letting me add you as a reviewer, or I would have done it.

### CODE REVIEW

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 1]
```diff
@@ -450,25 +448,28 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
                                                  dim=1,
                                                  index=Variable(linking_scores.data.new(row_type_index)).long())
 
-            # (num_question_tokens, num_entities_per_type)
+            # (num_question_tokens, num_entities_per_type + 1)
```
Comment (pdasigi) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Is the `+1` for the null entities? A comment would be nice. Also update above the lines with index_select`.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 2]
```diff
@@ -450,25 +448,28 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
...
             cell_type_scores[:, 0] = 0
             row_type_scores[:, 0] = 0
 
-            probabilities_cell = torch.nn.functional.softmax(cell_type_scores)
-            probabilities_row = torch.nn.functional.softmax(row_type_scores)
+            probabilities_cell = torch.nn.functional.softmax(cell_type_scores, dim=1)
+            probabilities_row = torch.nn.functional.softmax(row_type_scores, dim=1)
```
Comment (pdasigi) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Looks like these softmaxes need to be masked.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 3]
```diff
@@ -450,25 +448,28 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
...
+
+            # NOTE: The way that we're doing this here relies on the fact that entities are
+            # implicitly sorted by their types when we sort them by name, and that "fb:cell" comes
+            # before "fb:row".  This is not a great assumption, and could easily break later, but
+            # it should work for now.
+            to_cat = [probabilities_cell[:, 1:], probabilities_row[:, 1:]]
+            num_kept_entities = len(cell_type_index) + len(row_type_index) - 2
```
Comment (pdasigi) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: What is `-2`?

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 4]
```diff
@@ -450,25 +448,28 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
...
-            probabilities_cell = torch.nn.functional.softmax(cell_type_scores)
-            probabilities_row = torch.nn.functional.softmax(row_type_scores)
+            probabilities_cell = torch.nn.functional.softmax(cell_type_scores, dim=1)
+            probabilities_row = torch.nn.functional.softmax(row_type_scores, dim=1)
+
+            # NOTE: The way that we're doing this here relies on the fact that entities are
+            # implicitly sorted by their types when we sort them by name, and that "fb:cell" comes
+            # before "fb:row".  This is not a great assumption, and could easily break later, but
+            # it should work for now.
+            to_cat = [probabilities_cell[:, 1:], probabilities_row[:, 1:]]
```
Comment (rajasagashe) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: minor: maybe to_cat could be named probabilities_all_types and num_kept_entities to num_entities_in_batch

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 5]
```diff
@@ -428,9 +428,7 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
             Contains all the probabilities for an entity given a question word.
         """
         batch_size, num_question_tokens, num_entities = linking_scores.size()
-        batch_probabilities = Variable(linking_scores.data.new(torch.zeros(batch_size,
-                                                                           num_question_tokens,
-                                                                           num_entities)))
+        batch_probabilities = []
 
         for batch_index, world in enumerate(worlds):
             cell_type_index, row_type_index = [0], [0]
```
Comment (rajasagashe) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: A comment saying initialized with null entity index would be nice.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 3 above)
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: since each have the null entity, so this gets the number of non-null entities of both types

**Comment by rajasagashe** [comment]
Looks great! Very clean and concise.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 1 above)
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 2 above)
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: They don't, actually, because we constructed a `cell_type_scores` matrix per batch instance, with no padding.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 4 above)
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Sounds good.
